components/scream/src/control/atmosphere_driver.cpp

```diff
@@ -181,14 +186,64 @@ void AtmosphereDriver::finalize ( /* inputs? */ ) {
 #endif
 }
 
+void AtmosphereDriver::register_groups () {
+  using ci_string = ekat::CaseInsensitiveString;
+  using ci_string_pair = std::pair<ci_string,ci_string>;
+
+  // Given a list of group-grid pairs (A,B), make sure there is a copy of A
+  // on grid B registered in the repo.
+  auto lambda = [&](const std::set<ci_string_pair>& groups_grids) {
+    const auto& groups_to_fnames = m_field_repo->get_field_groups_names();
+    for (const auto& gg : groups_grids) {
+      const auto& group = gg.first;
+      const auto& grid  = gg.second;
+      EKAT_REQUIRE_MSG(groups_to_fnames.find(group)!=groups_to_fnames.end(),
+        "Error! Group '" << group << "' not found in the repo.\n");
+
+      const auto& fnames = groups_to_fnames.at(group);
+      for (const auto& name : fnames) {
+        auto aliases = m_field_repo->get_aliases(name);
+        EKAT_REQUIRE_MSG(aliases.size()>0,
+          "Error! Something went wrong while looking for field '" << name << "' in the repo.\n");
+
+        // Check if a copy of this field on the right grid already is registered.
+        bool found = false;
+        for (const auto& alias : aliases) {
+          if (alias.first.get_grid_name()==grid) {
+            found = true;
+            break;
+          }
+        }
+
+        if (!found) {
+          // Field $name in group $group has no copy on grid $grid.
+          // Lets take any fid in the repo for this field, and register
+          // a copy of it on grid $grid. We can do this by creating
+          // a remapper and using its capabilities.
+          const auto& alias = *aliases.begin();
+          const auto& fid = alias.first;
+          auto r = m_grids_manager->create_remapper(fid.get_grid_name(),grid);
+          auto f_units = fid.get_units();
+          auto src_layout = fid.get_layout();
+          auto tgt_layout = r->create_tgt_layout(src_layout);
+          FieldIdentifier new_fid(name,tgt_layout,f_units,grid);
+          m_field_repo->register_field(new_fid,group);
+        }
+      }
+    }
+  };
+  lambda( m_atm_process_group->get_required_groups() );
```

Comment:
Any danger that this data might be needed before these lambdas finish? I'm guessing not, but the pedant in me has to ask.

Reply:
You mean danger that data is erased before lambda finishes? I don't think so. I just used a lambda cause I didn't want to create a method for the AD. I'm lazy.

Do you think it's best to add a method?

Reply:
No, I don't think it matters.

Reply:
If we want to be ultra paranoid about inconsistency, we can always add a barrier after the lambda. But I think it's unlikely that we need to worry about this.

Reply:
There should be no threading occurring in the lambda body, so yeah, this should be safe.